---
title: "Unleash NBNB on 10x PBMC data"
output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    toc_float: true
    self_contained: true
    mathjax: default
    code_download: true
---

# Script setup

## Set CPU / Threads
```{r}
R.utils::setOption("mc.cores", 8)
library(pbmcapply)
```


## Load data from SDS:
```{r}
# file lie on the SDS:
sds10x <- file.path("~","sds","sd17l002","p","scRNAseq_datasets",
                    "10xGenomics_pbmc3k", "filtered_gene_bc_matrices", "hg19") 

rawC <- Seurat::Read10X(sds10x)
sf <- Matrix::colSums(rawC) / mean(Matrix::colSums(rawC))
```

## Function: fitNB (R)
```{r}
disp <- function(k) (var(k) - mean(k)) / mean(k) / mean(k) # var = mu + disp * mu^2
  # remember to divide k by sf, and to prohibit negative values

fitNB <- function(k, sf, initialVals = c(mean(k), disp(k))) {   
    o = optim( 
      initialVals,  # mu, alpha
      function(x) {
        -sum( lgamma( k + 1/x[2] ) - lgamma( 1/x[2] ) - lgamma( k+1 ) - ( k + 1/x[2] ) * log( 1 + sf * x[2] * x[1] ) + k * log( sf * x[2] * x[1] ) )
        },
      
      function(x) c(
        -sum( ( k - sf * x[1] ) / ( x[1] + sf * x[2] * x[1]^2 ) ),
        -sum( ( x[2] * ( k - sf * x[1] ) / ( 1 + sf * x[2] * x[1] ) + log( 1 + sf * x[2] * x[1] ) - digamma( k + 1/x[2] ) + digamma( 1/x[2] ) ) / x[2]^2 ) ),
      lower = c( 1e-10, 1e-10 ),
      method = "L-BFGS-B" )
    c( mean = o$par[1], disp = o$par[2] ) }


library(purrr) # for possibly
safe_fitNB <- possibly(fitNB, otherwise = c(mean=NA, disp=NA))

```

# Start values matter

Although using the `disp` function is slower, I would trust the estimated 
dispersion more with it for low mus:
```{r}
g <- "TIE1"
  microbenchmark::microbenchmark(
  safe_fitNB(rawC[g, ], sf, c(mean(rawC[g,]), 1)),
  safe_fitNB(rawC[g, ], sf, c(mean(rawC[g, ]), disp(rawC[g, ]))),
  times = 20
  )
  
  safe_fitNB(rawC[g, ], sf, c(mean(rawC[g,]), 1))
  safe_fitNB(rawC[g, ], sf, c(mean(rawC[g, ]), disp(rawC[g, ])))
```

# Find super-poissonian genes

```{r}
disptable <- pbmclapply(rownames(rawC), function(g) {
   NBparams <- safe_fitNB(rawC[g, ], sf)
   data.frame(gene = g, mean = NBparams[1], disp = NBparams[2], row.names = NULL)
 })

# saveRDS(do.call(rbind, disptable), "/home/felix/sc_methods/nb_sc/Robj_superpoissonGenes.rds")
```




